src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -103,6 +103,133 @@ void CodeGen::genSSEIntrinsic(GenTreeHWIntrinsic* node)
             op2Reg = op2->gtRegNum;
             emit->emitIns_SIMD_R_R_R(INS_addps, targetReg, op1Reg, op2Reg, TYP_SIMD16);
             break;
+
+        case NI_SSE_And:
```

Comment:
All the cases (so far) are essentially the same with only the instruction (or immediate value) differing.

It might be worthwhile including this info in the `hwintrinsiclist` and having them be entries in an array for constant time lookup...

Reply:
Yes, unlike "classic" instructions SSE/AVX instructions have a more consistent format so it should possible to drive most codegen from tables. Basically you need to add to the list a "form" value that can be used to select between `R_R_R` and `R_R_R_I` and of course, the instruction itself.

Well, there's also the problem of memory operands but that's something that the IR node will indicate. So `R_R_R` in the hw intrinsic list really stands for `R_R_RM`.

Reply:
I was actually just thinking of having a single array, something where the entries are {instruction, immediate}. Helper functions (like `get_IsSupported`) would have `INS_INVALID` and instructions without an immediate would probably have `-1`.

There can then be two general groupings. One for instructions without an immediate (those that use `SIMD_R_R_R`) and one for instructions with an immediate (those that use `SIMD_R_R_R_I`).

Other instructions (such as `Load`) which have different semantics can be in their own case grouping.

I don't think we necessarily need a `form` value, unless you think that would be more efficient.

Reply:
I do not want to tabulate the codegen of hw intrinsic, which won't always be such simple. 
For example, `intrinsicID` + `base` is not enough to indicate codegen for some intrinsic, I would add a new field into `GenTreeHWIntrinsic` to represent overloads that would be used here.

Reply:
> I don't think we necessarily need a form value, unless you think that would be more efficient.

The `form` thing was just a suggestion. Any solution that cuts down the amount of repeated code should be good enough, provided that it doesn't cause other problem.

>  I would add a new field into GenTreeHWIntrinsic to represent overloads that would be used here.

Hmm? It's not clear why that would be better.

Reply:
I don't think we should do it for all instructions, but a number of them (especially the binary arithmetic and comparison operators) will all be fairly common and follow roughly the same code path.

Cutting down duplicate code will likely be worthwhile for these operations.

Reply:
> Hmm? It's not clear why that would be better.

@mikedn For example, we need an additional field to distinghuish these two intrinsics (`intrinsicID` + `baseType` is not enough).
```csharp
Vector128<float> GatherVector128(float* baseAddress, Vector128<long> index, byte scale);
Vector128<float> GatherVector128(float* baseAddress, Vector128<int> index, byte scale);
```
My current is to add a `overloads` field in `GenTreeHWIntrinsic` to bring that information from importer to CodeGen.

Reply:
Here is the complete list of `GatherVector128` to show the necessity.
```CSharp
        /// <summary>
        /// __m128i _mm_i32gather_epi32 (int const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i32gather_epi32 (int const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<uint> GatherVector128(uint* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i32gather_epi64 (__int64 const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<long> GatherVector128(long* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i32gather_epi64 (__int64 const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<ulong> GatherVector128(ulong* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128 _mm_i32gather_ps (float const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<float> GatherVector128(float* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128d _mm_i32gather_pd (double const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<double> GatherVector128(double* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i64gather_epi32 (int const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i64gather_epi32 (int const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<uint> GatherVector128(uint* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i64gather_epi64 (__int64 const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<long> GatherVector128(long* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i64gather_epi64 (__int64 const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<ulong> GatherVector128(ulong* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128 _mm_i64gather_ps (float const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<float> GatherVector128(float* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128d _mm_i64gather_pd (double const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<double> GatherVector128(double* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
```

Reply:
> intrinsicID + baseType is not enough

Hmm, can't we use a different intrinsic id for each overload?

Reply:
> Hmm, can't we use a different intrinsic id for each overload?

That will break the intrinsic searching because a method name has to map to one intrinsic id in the current `[Intrinsic]` system.